packages/backend/src/shared/pagination.ts

```diff
@@ -0,0 +1,114 @@
+import { Asset } from '../asset/model'
+import { AssetService } from '../asset/service'
+import { IncomingPayment } from '../open_payments/payment/incoming/model'
+import { IncomingPaymentService } from '../open_payments/payment/incoming/service'
+import { OutgoingPayment } from '../open_payments/payment/outgoing/model'
+import { OutgoingPaymentService } from '../open_payments/payment/outgoing/service'
+import { Quote } from '../open_payments/quote/model'
+import { QuoteService } from '../open_payments/quote/service'
+import { Peer } from '../peer/model'
+import { PeerService } from '../peer/service'
+import { PageInfo, Pagination } from './baseModel'
+
+type ListResult = {
+  pagination: PageInfo
+  result: unknown // (IncomingPaymentJSON | OutgoingPaymentJSON | QuoteJSON)[]
+}
+
+export async function list(
+  service: IncomingPaymentService | OutgoingPaymentService | QuoteService,
+  host: string,
+  accountId: string,
+  pagination?: Pagination
+): Promise<ListResult> {
+  const page = await service.getAccountPage(accountId, pagination)
+  const pageInfo = await getPageInfo(service, page)
+  if (pagination?.last) {
+    pageInfo.last = page.length
+  } else {
+    pageInfo.first = page.length
+  }
+  return {
+    pagination: pageInfo,
+    result: page.map((item: IncomingPayment | OutgoingPayment | Quote) => {
+      return {
+        ...item.toJSON(),
+        accountId: `${host}/${accountId}`,
+        id:
+          item['receivedAmount'] !== undefined
+            ? `${host}/${accountId}/incoming-payments/${item.id}`
+            : item['sentAmount'] !== undefined
+            ? `${host}/${accountId}/outgoing-payments/${item.id}`
+            : `${host}/${accountId}/quotes/${item.id}`
+      }
+    })
+  }
+}
+
+export async function getPageInfo(
+  service:
+    | IncomingPaymentService
+    | OutgoingPaymentService
+    | QuoteService
+    | AssetService
+    | PeerService,
+  page: (IncomingPayment | OutgoingPayment | Quote | Asset | Peer)[],
+  accountId?: string
```

Comment:
```suggestion
export async function getPageInfo<T extends BaseModel>(
  getPage: (pagination: Pagination) => Promise<T[]>,
  page: T[],
```

`getAccountPage` can be passed to `getPage` as:
```
const pageInfo = await getPageInfo((pagination) => quoteService.getAccountPage(accountId, pagination), page)
```

Reply:
:exploding_head: I knew my solution wasn't great and I'm blown away by the elegance of this. Thank you!